graphjet-core/src/main/java/com/twitter/graphjet/algorithms/MultiThreadedPageRank.java

```diff
@@ -0,0 +1,237 @@
+    prVector = new AtomicDoubleArray((int) (maxNodeId + 1));
+    nodes.forEach(v -> prVector.set((int) (long) v, 1.0 / nodeCount));
+
+    // We're going to divide the nodes into partitions, one for each thread.
+    LongArrayList[] nodePartitions = new LongArrayList[threads];
+    int partitionSize = nodes.size() / threads;
+    for (int i=0; i<threads; i++) {
+      int startPos = i * partitionSize;
+      // The final partition get the rest of the nodes.
+      int endPos = i == (threads - 1) ? nodes.size() : (i + 1) * partitionSize;
+      nodePartitions[i] = new LongArrayList(nodes.subList(startPos, endPos));
+    }
+
+    int i = 0;
+    while (i < this.maxIterations && normL1 > tolerance) {
+      iterate(dampingAmount, noOuts, nodePartitions);
+      i++;
+    }
+
+    return i;
+  }
+
+  /**
+   * Returns the final L1 norm value after PageRank has been run.
+   *
+   * @return the final L1 norm value after PageRank has been run
+   */
+  public double getL1Norm() {
+    return normL1;
+  }
+
+  /**
+   * Returns the PageRank vector, or null if PageRank has not yet been run.
+   *
+   * @return the PageRank vector, or null if PageRank has not yet been run
+   */
+  public AtomicDoubleArray getPageRankVector() {
+    return prVector;
+  }
+
+  /**
+   * A PageRank worker thread that distributes PageRank mass for a partition of nodes in a
+   * particular iteration.
+   */
+  private class PageRankWorker extends Thread {
+    final int id;
+    final AtomicDoubleArray nextPR;
+    final LongArrayList nodes;
+    final CountDownLatch latch;
+    final double mass;
+
+    /**
+     * Creates a PageRank worker thread.
+     *
+     * @param id      partition id
+     * @param nextPR  the PageRank vector to modify
+     * @param nodes   the nodes this thread is responsible for
+     * @param latch   countdown latch to synchronize all worker threads for an iteration
+     * @param mass    PageRank mass to pass along (from dangling nodes and from damping)
+     */
+    public PageRankWorker(int id, AtomicDoubleArray nextPR, LongArrayList nodes,
+                          CountDownLatch latch, double mass) {
+      this.id = id;
+      this.nextPR = nextPR;
+      this.nodes = nodes;
+      this.latch = latch;
+      this.mass = mass;
+    }
+
+    @Override
+    public void run() {
+      // Each thread runs on its own partition of the nodes, distributing PageRank mass for the
+      // nodes that were assigned to it.
+      final LongIterator iter = nodes.iterator();
+      while (iter.hasNext()) {
+        long v = iter.nextLong();
+        int outDegree = graph.getOutDegree(v);
+        if (outDegree > 0) {
+          double outWeight = dampingFactor * prVector.get((int) v) / outDegree;
+          EdgeIterator edges = graph.getOutEdges(v);
+          while (edges.hasNext()) {
+            int nbr = (int) edges.nextLong();
+            // Note that getAndAdd is implemented as an atomic operation with CAS, so
```

Comment:
[i'm hallucinating that i commented on this before, but trying again]

sorry, i'm not sure i understand this bit --- CAS will ensure that if multiple threads try to update the nbr weight then only one will succeed, but others will fail. since you move on without retrying, not all weight will be transferred to the node in that iteration. thus, it is not a true pagerank iteration, right?